### PR TITLE
Open code the __fastfail intrinsic for rtabort! on windows

### PR DESCRIPTION
**Pull request by lambda**
As described https://msdn.microsoft.com/en-us/library/dn774154.aspx

This is a Windows 8+ mechanism for terminating the process quickly,
which degrades to either an access violation or bugcheck in older versions.

I'm not sure this is better the the current mechanism of terminating
with an illegal instruction, but we recently converted unix to
terminate more correctly with SIGABORT, and this _seems_ more correct
for windows.

[breaking-change]

### CODE REVIEW

**Comment by lambda** [comment]
This is essentially a rebase of #31519 from @brson on top of the latest version of #31457 which was merged.

r? @alexcrichton

**Comment by lambda** [comment]
cc @retep998

File: src/libstd/sys/common/util.rs [hunk 1]
```diff
@@ -46,11 +46,24 @@ unsafe fn abort_internal() -> ! {
     libc::abort()
 }
 
-// On Windows, we want to avoid using libc, and there isn't a direct
-// equivalent of libc::abort.  The __failfast intrinsic may be a reasonable
-// substitute, but desireability of using it over the abort instrinsic is
-// debateable; see https://github.com/rust-lang/rust/pull/31519 for details.
-#[cfg(not(unix))]
+// On windows, use the processor-specific __fastfail mechanism
+// https://msdn.microsoft.com/en-us/library/dn774154.aspx
```
Comment (alexcrichton) [inline src/libstd/sys/common/util.rs]: Could this elaborate on why we're using this? For example mention that exception handlers are removed and such, and also that on Windows 7 and prior we do *not* see this behavior but it largely equates to the same semantics.

File: src/libstd/sys/common/util.rs [hunk 2]
```diff
@@ -46,11 +46,24 @@ unsafe fn abort_internal() -> ! {
...
+#[cfg(all(windows, target_arch = "x86_64"))]
+unsafe fn abort_internal() -> ! {
+    asm!("int $$0x29" :: "{rcx}"(7) ::: volatile); // 7 is FAST_FAIL_FATAL_APP_EXIT
+    ::intrinsics::unreachable();
+}
+
+// On other platforms, just use the LLVM intrinsic.
+#[cfg(not(any(unix,
+              all(windows, target_arch = "x86"),
+              all(windows, target_arch = "x86_64"))))]
```
Comment (alexcrichton) [inline src/libstd/sys/common/util.rs]: It's ok to just remove this fallback for now, it'd just be here for non-intel Windows which would needs lots of changes to support anyway. That, and we'd also want to come back and update this if we ever added support there as well.

File: src/libstd/sys/common/util.rs [hunk 3]
```diff
@@ -46,11 +46,24 @@ unsafe fn abort_internal() -> ! {
...
+// https://msdn.microsoft.com/en-us/library/dn774154.aspx
+#[cfg(all(windows, target_arch = "x86"))]
+unsafe fn abort_internal() -> ! {
+    asm!("int $$0x29" :: "{ecx}"(7) ::: volatile); // 7 is FAST_FAIL_FATAL_APP_EXIT
+    ::intrinsics::unreachable();
+}
+
+#[cfg(all(windows, target_arch = "x86_64"))]
+unsafe fn abort_internal() -> ! {
+    asm!("int $$0x29" :: "{rcx}"(7) ::: volatile); // 7 is FAST_FAIL_FATAL_APP_EXIT
```
Comment (retep998) [inline src/libstd/sys/common/util.rs]: Technically speaking, on x86_64 if you move 7 into ecx, it'll automatically zero extend it to rcx, so you can use a single ecx implementation for both x86 and x86_64.

File: src/libstd/sys/common/util.rs (hunk 2 above)
Comment (lambda) [inline src/libstd/sys/common/util.rs]: That makes my inner exhaustiveness checker cry out in pain!

Comment (alexcrichton) [inline src/libstd/sys/common/util.rs]: On the other hand though there's a documented way to do this for ARM Windows and if we ever add support for that then it's arguably a bug *unless* we update this, so this'll be a good notification to do that. We could always add this fallback implementation in that case as well if it's the most appropriate at the time.

Comment (lambda) [inline src/libstd/sys/common/util.rs]: Good point. Removed the fallback for now, we can always add it back again if there are any platforms that aren't properly covered.

File: src/libstd/sys/common/util.rs (hunk 1 above)
Comment (lambda) [inline src/libstd/sys/common/util.rs]: Added an expanded explanation.

File: src/libstd/sys/common/util.rs (hunk 3 above)
Comment (lambda) [inline src/libstd/sys/common/util.rs]: Sure, collapsed the two implementations based on this suggestion.

**Comment by alexcrichton** [comment]
@bors: r+ 696a570a00db2528b7934be32da7d8334738a997

Thanks!

**Comment by durka** [comment]
What does "open code" mean, btw? Is it a synonym for "hard code"?

**Comment by lambda** [comment]
Open-code is compiler terminology for directly generating the instructions to perform a primitive operation rather than calling a function to do so. See [this exercise in SICP for example](https://mitpress.mit.edu/sicp/full-text/book/book-Z-H-35.html#%_thm_5.38). I think "inline the assembly" might be a better translation than "hard code".

**Comment by retep998** [comment]
Again with the spurious failure due to arguments to Command being too long.

@alexcrichton Should I submit a PR that makes rustc print out the linker arguments whenever it hits error 206 on Windows?

**Comment by alexcrichton** [comment]
@retep998 sure!

@bors: retry

**Comment by retep998** [comment]
Test failure is https://github.com/rust-lang/rust/issues/33778

**Comment by pravic** [comment]
> User-mode fast fail requests appear as a second chance **non-continuable exception**

> Windows operating systems that do not support the fast fail instruction natively will typically treat a fast fail request as **an access violation**, or as an UNEXPECTED_KERNEL_MODE_TRAP bugcheck. In these cases, the program is still terminated, but not necessarily as quickly.

In Windows 7 this is just a regular **continuable** exception, so it can be catched, suppressed or recovered (for example, by SEH/VEH directly or undirectly).

**Comment by retep998** [comment]
@pravic The previous solution of generating an illegal instruction exception was also a regular continuable exception, so there is no regression here on Windows 7, but there is an improvement on newer versions.

**Comment by pravic** [comment]
So why not to raise a noncontinuable exception? Or no. Why not to reuse the ucrt behavior? Looks hastily.

**Comment by retep998** [comment]
@pravic By all means feel free to submit a PR that raises a noncontinuable exception on Windows 7 and older. This PR does not prevent such a change being made later on, and is simply a step in the right direction.

**Pull request merged**